tests/internal/aws_client_mock.c

```diff
@@ -0,0 +1,250 @@
+    memset(mock->surrogate, 0, sizeof(struct flb_aws_client));
+
+    /* Switch vtable to mock vtable */
+    mock->super.client_vtable = &mock_client_vtable;
+    mock->request_chain = request_chain;
+    mock->next_request_index = 0;
+    return mock;
+}
+
+/* Destroy flb_aws_client_mock */
+void flb_aws_client_mock_destroy(struct flb_aws_client_mock *mock)
+{
+    /* Remove from generator registry if stored */
+    if (flb_aws_client_mock_instance == mock) {
+        flb_aws_client_mock_instance = NULL;
+    }
+
+    /* Resurrect surrogate, and destroy flb_aws_client */
+    *mock->surrogate = mock->super;
+    flb_aws_client_destroy(mock->surrogate);
+
+    /* Destroy mock flb_aws_client */
+    flb_free(mock);
+}
+
+/* Return a Mocked flb_aws_client, ready for injection */
+struct flb_aws_client *flb_aws_client_mock_context(struct flb_aws_client_mock *mock)
+{
+    return (struct flb_aws_client *)mock;
+}
+
+/* Get the number of unused requests */
+int flb_aws_client_mock_count_unused_requests(struct flb_aws_client_mock *mock)
+{
+    return mock->request_chain->length - mock->next_request_index;
+}
+
+/* Set flb_aws_client_mock_instance used in mock generator */
+void flb_aws_client_mock_set_generator_instance(struct flb_aws_client_mock *mock)
+{
+    flb_aws_client_mock_instance = mock;
+}
+
+/* Set flb_aws_client_mock_instance used in mock generator */
+struct flb_aws_client_mock *flb_aws_client_mock_get_generator_instance(
+    struct flb_aws_client_mock *mock)
+{
+    return flb_aws_client_mock_instance = mock;
+}
+
+/* Get generator used in mock */
+struct flb_aws_client_generator *flb_aws_client_get_mock_generator()
+{
+    return &mock_generator;
+}
+
+/* Get the number of unused requests */
+int flb_aws_client_mock_generator_count_unused_requests()
+{
+    TEST_ASSERT(flb_aws_client_mock_instance != 0);
+    return flb_aws_client_mock_count_unused_requests(flb_aws_client_mock_instance);
+}
+
+/* Return the mock instance */
+struct flb_aws_client *flb_aws_client_create_mock()
+{
+    TEST_CHECK(flb_aws_client_mock_instance != NULL);
+    TEST_MSG(
+        "[aws_mock_client] Must initialize flb_aws_client_mock_instance before calling "
+        "flb_aws_client_create_mock()");
+    TEST_MSG(
+        "[aws_mock_client] This ouccurs when the generator is called, before tests are "
+        "initialized.");
+
+    return flb_aws_client_mock_context(flb_aws_client_mock_instance);
+}
+
+/* Mock request used by flb_aws_client mock */
+static struct flb_http_client *flb_aws_client_mock_vtable_request(
+    struct flb_aws_client *aws_client, int method, const char *uri, const char *body,
+    size_t body_len, struct flb_aws_header *dynamic_headers, size_t dynamic_headers_len)
+{
+    /* Get access to mock */
+    struct flb_aws_client_mock *mock = (struct flb_aws_client_mock *)aws_client;
+
+    /* Check that a response is left in the chain */
+    if (!TEST_CHECK(mock->next_request_index < mock->request_chain->length)) {
```

Comment:
nit here and other places: while I am fine with this code style, strictly speaking fluent bit prefers you do not call functions in an if condition. Call it outside and then use the return value in an if block. This is supposed to be more readable I guess.

Reply:
This sounds good. Will update.